### PR TITLE
refactor: Batch write trait

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

To take into account the fact that the size we get initially in the iroh-bytes protocol is not verified, I want to lazily move from mem storage to file storage only once some limit is *actually* exceeded.

This requires combining outboard and data file and treating them a bit more as one file to which you write batches.

For now, just add the trait and keep the impl the same.

## Notes & open questions

Should the BaoBatchWriter move to bao-tree io? I think in the long term yes, but for now let's keep it here to make things simple?

Note: this kind of code *really* wants 1.75 impl return types in traits.... I asked on the public discord if somebody has some major objections, otherwise I would just go with it.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

File: iroh-bytes/src/get.rs [hunk 1]
```diff
@@ -491,6 +494,16 @@ pub mod fsm {
             Ok(res)
         }
 
+        /// Write the entire stream for this blob to a batch writer.
+        pub async fn write_all_batch<B>(self, batch: B) -> result::Result<AtEndBlob, DecodeError>
+        where
+            B: BaoBatchWriter,
+        {
+            let (content, _size) = self.next().await?;
```
Comment (dignifiedquire) [inline iroh-bytes/src/get.rs:502]: would it make sense to report the total written size in return value?

File: iroh-bytes/src/store/traits.rs [hunk 2]
```diff
@@ -103,23 +110,178 @@ pub trait Map: Clone + Send + Sync + 'static {
...
+/// to the implementation.
+pub trait BaoBatchWriter {
+    /// Write a batch of bao content items to the underlying storage.
+    ///
+    /// The batch is guaranteed to be sorted as data is received from the network.
+    /// So leafs will be sorted by offset, and parents will be sorted by pre order
+    /// traversal offset. There is no guarantee that they will be consecutive
+    /// though.
+    ///
+    /// The size is the size the remote side told us. It is not guaranteed
```
Comment (dignifiedquire) [inline iroh-bytes/src/store/traits.rs:130]: this is the total size of all batch items?

File: iroh-bytes/src/store/traits.rs [hunk 3]
```diff
@@ -103,23 +110,178 @@ pub trait Map: Clone + Send + Sync + 'static {
...
+    /// necessary in typical io code.
+    ///
+    /// TODO: return impl Future<...> once we got 1.75
+    fn write_batch(
+        &mut self,
+        size: u64,
+        batch: Vec<BaoContentItem>,
+    ) -> LocalBoxFuture<'_, io::Result<()>>;
+
+    /// Sync the writer
```
Comment (dignifiedquire) [inline iroh-bytes/src/store/traits.rs:147]: maybe specify to what this syncs, like that it persists or so, and not synced to another node ;)

File: iroh-bytes/src/store/traits.rs [hunk 4]
```diff
@@ -103,23 +110,178 @@ pub trait Map: Clone + Send + Sync + 'static {
...
+            }
+            Ok(())
+        }
+        .boxed_local()
+    }
+
+    fn sync(&mut self) -> LocalBoxFuture<'_, io::Result<()>> {
+        async move {
+            self.data.sync().await?;
+            self.outboard.sync().await?;
```
Comment (dignifiedquire) [inline iroh-bytes/src/store/traits.rs:268]: could these be polled using `join`?

File: iroh-bytes/src/store/traits.rs [hunk 5]
```diff
@@ -103,23 +110,178 @@ pub trait Map: Clone + Send + Sync + 'static {
...
+}
+
+impl<D, O> BaoBatchWriter for CombinedBatchWriter<D, O>
+where
+    D: AsyncSliceWriter,
+    O: OutboardMut,
+{
+    fn write_batch(
+        &mut self,
+        _size: u64,
```
Comment (dignifiedquire) [inline iroh-bytes/src/store/traits.rs:246]: why is the size not needed in this impl?

File: iroh-bytes/src/store/traits.rs (hunk 2 above)
Comment (rklaehn) [reply iroh-bytes/src/store/traits.rs:130]: No, this is the size in bytes of the entire blob. We just can't fully trust it yet. But in the previous impl we did trust it.

This determines what the stuff on the wire means, so this is an *upper bound* on the size, so are guaranteed to only get stuff within 0..size.

Comment (rklaehn) [reply iroh-bytes/src/store/traits.rs:130]: Trying to clarify this more in the comments.

Comment (dignifiedquire) [reply iroh-bytes/src/store/traits.rs:130]: maybe call it blob_size or sth like that

File: iroh-bytes/src/get.rs (hunk 1 above)
Comment (rklaehn) [reply iroh-bytes/src/get.rs:502]: This returns a state of the iroh-bytes state machine. If you follow the state machine to the end without an error, you get a Stats object which contains this information.

But it is hard to come up with stats that are not too expensive to compute and that are useful for everybody. So to get very precise info you would have to write a wrapper for the batch writer, just like the FallibleProgressBatchWriter, that computes some detailed stats.

```
pub struct Stats {
    /// The number of bytes written
    pub bytes_written: u64,
    /// The number of bytes read
    pub bytes_read: u64,
    /// The time it took to transfer the data
    pub elapsed: Duration,
}
```

File: iroh-bytes/src/store/traits.rs (hunk 4 above)
Comment (rklaehn) [reply iroh-bytes/src/store/traits.rs:268]: yeah, but this thing is temporary anyway. The idea of the batch writer in the long term is to be able to batch things including sync, and to leave it up to the impl what needs to be synced.

So the long term impl would be spawn_blocking(|| { sync all the files }) or something.

File: iroh-bytes/src/store/traits.rs (hunk 5 above)
Comment (rklaehn) [reply iroh-bytes/src/store/traits.rs:246]: We are not quite there yet in terms of tolerating wrong sizes.

The OutboardMut has a Tree, which has a size. So the OutboardMut needs to know in advance what the size is. It gets this info from the PartialEntry, which in this PR still has a size.

The fun part will be one of the next PRs, where we start using this size field here.

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**